### PR TITLE
Opening hours reworking

### LINKED ISSUE #4360: Opening hours for galleries/library/cafe/restaurant/shop
Follow on work from https://github.com/wellcometrust/wellcomecollection.org/issues/3439#issuecomment-478541329

## Opening times page
- A new component to show opening hours:
Regular: list all hours per venue
Holiday hours: list all hours for the duration of the holiday (display two weeks prior)
Easter weekend (Friday-Sunday), Bank holidays (Saturday-Monday), Christmas & New year's (Christmas Eve–New Year's day)
- Library closures
list all closures dates for the year in advance
- Busy times 
Content card links to new 'Busy times' page

## Additional info
- Serving times for Wellcome Kitchen and Wellcome Café are not part of the component but are free text on the Kitchen and Café pages.
- Decided not to display the holiday messaging in the info banner to put less emphasis on physical building. We will explore a 'Visit us' card in the landing pages in future.

## Where to display
 All display on https://wellcomecollection.org/opening-times
* Library component also displays on https://wellcomecollection.org/pages/Wuw19yIAAK1Z3Smm
* Cafe component also displays on https://wellcomecollection.org/pages/Wvl1wiAAADMJ3zNe
* Restaurant component also displays on https://wellcomecollection.org/pages/Wuw19yIAAK1Z3Snk
* Shop component also displays on https://wellcomecollection.org/pages/WwgaIh8AAB8AGhC_
* Galleries and Reading Room component only displays on Opening times page

## Acceptance Criteria

- [x] All display on https://wellcomecollection.org/opening-times
- [ ] Library component also displays on https://wellcomecollection.org/pages/Wuw19yIAAK1Z3Smm
- [ ] Cafe component also displays on https://wellcomecollection.org/pages/Wvl1wiAAADMJ3zNe
- [ ] Restaurant component also displays on https://wellcomecollection.org/pages/Wuw19yIAAK1Z3Snk
- [ ] Shop component also displays on https://wellcomecollection.org/pages/WwgaIh8AAB8AGhC_
- [ ] Galleries and Reading Room component only displays on Opening times

### PR DESCRIPTION
**Pull request by davidpmccormick**
☝️ 

Closes #4360 and #4361 

__Today's hours in the footer__:
Screenshot 2019-04-11 at 16 17 17

__Opening times page__ (image, heading = {venue name}, link to venue page:
Screenshot 2019-04-11 at 16 22 31

__Venue page__ (no image, heading = 'Opening hours', link to opening times page):
Screenshot 2019-04-11 at 16 18 25

### CODE REVIEW

File: common/model/opening-hours.js [hunk 1]
```diff
@@ -22,20 +22,20 @@ export type ExceptionalPeriod = {|
 
 export type OpeningHoursDay = {|
   dayOfWeek: Day,
-  opens?: string,
-  closes?: string,
+  opens?: ?string,
```
Comment (jamesgorrie) [inline common/model/opening-hours.js]: Always feels a bit weird to have a nullable + optional property?
Is it not one or the other? e.g. nullable, but you have to send `null` if it is?

File: common/services/prismic/opening-times.js [hunk 2]
```diff
@@ -108,211 +94,185 @@ function exceptionalOpeningPeriods(dates: OverrideDate[]): ExceptionalPeriod[] {
...
-
-  return exceptionalPeriods.length > 0 ? exceptionalPeriods : null;
+export function groupConsecutiveDays(
+  openingHoursPeriods: ExceptionalOpeningHoursDay[][]
+): ExceptionalOpeningHoursDay[][] {
+  const flattenedArray = openingHoursPeriods.reduce(
+    (acc, curr) => acc.concat(curr),
+    []
+  );
+  return flattenedArray.length > 0
```
Comment (jamesgorrie) [inline common/services/prismic/opening-times.js]: This conditional doesn't make sense, as if the `length` is 0, running the `sort / reduce` would just return an array any way?

File: common/views/components/Body/Body.js [hunk 3]
```diff
@@ -165,6 +168,15 @@ const Body = ({ body, isDropCapped }: Props) => {
               </Layout10>
             )}
 
+            {slice.type === 'collectionVenue' && (
+              <VenueHours
+                venue={parseVenueTimesToOpeningHours(slice.value)}
+                isInList={
+                  pageId === 'openingTimes' || pageId === 'WwQHTSAAANBfDYXU'
```
Comment (jamesgorrie) [inline common/views/components/Body/Body.js]: What's going on here?

File: common/views/components/Body/Body.js [hunk 4]
```diff
@@ -35,9 +37,10 @@ export type BodyType = BodySlice[];
 type Props = {|
   body: BodyType,
   isDropCapped?: boolean,
+  pageId: string,
```
Comment (jamesgorrie) [inline common/views/components/Body/Body.js]: Is a better name for this `contentId`?

File: prismic-model/json/pages.json [hunk 5]
```diff
@@ -279,9 +279,9 @@
               }
             }
           },
-          "venueTimes": {
+          "collectionVenue": {
             "type": "Slice",
-            "fieldset": "Venue opening times",
+            "fieldset": "Collection venue",
```
Comment (jamesgorrie) [inline prismic-model/json/pages.json]: It doesn't look like the JSON has been updated?
And are we sure these were not used anywhere?

File: content/webapp/pages/page.js [hunk 6]
```diff
@@ -105,7 +108,20 @@ export class Page extends Component<Props> {
...
         imageAltText={page.image && page.image.alt}
       >
-        <ContentPage id={page.id} Header={Header} Body={<Body body={body} />} />
+        <ContentPage
+          id={page.id}
+          Header={Header}
+          Body={<Body body={body} pageId={page.id} />}
+        />
+
+        {/* Display closures on the library page */}
```
Comment (jamesgorrie) [inline content/webapp/pages/page.js]: Quite a lot of brittle bespoke code for things that future selves will probably not like.

Could this be split up into smaller task, potentially looking at these pieces individually and maybe coming up with a more scalable solution?

File: common/views/components/Body/Body.js (hunk 3 above)
Comment (davidpmccormick) [reply common/views/components/Body/Body.js]: The `VenueHours` component behaves slightly differently if it's on the opening-times page, or on a venue page.

Although we have a vanity url for `opening-times`, the opening times page content comes from Prismic and it available at `/pages/WwQHTSAAANBfDYXU`, so makes sense to check for either.

File: common/views/components/Body/Body.js (hunk 4 above)
Comment (davidpmccormick) [reply common/views/components/Body/Body.js]: Maybe. Don't know.

File: prismic-model/json/pages.json (hunk 5 above)
Comment (gestchild) [reply prismic-model/json/pages.json]: - Do you mean the json for all the other content types that share Body?
- Yeah, this slice was created for this piece of work, so nothing else uses it

File: content/webapp/pages/page.js (hunk 6 above)
Comment (davidpmccormick) [reply content/webapp/pages/page.js]: We weren't totally in love with this particular shim, but seemed like the least esoteric of the options available:

1. Singleton type slice
2. HTML serialiser
3. hardcode in the template

Have we overlooked another option?

**Comment by davidpmccormick** [comment]
> Finding it quite hard to review as it's a big one

Yeah, apologies for that. FWIW @gestchild and I were pairing on most of it.

**Comment by jamesgorrie** [comment]
> FWIW @gestchild and I were pairing on most of it.

It's worth a lot, just mentioned as someone not directly involved it's hard.

File: prismic-model/json/pages.json (hunk 5 above)
Comment (jamesgorrie) [reply prismic-model/json/pages.json]: The point of the body though was to be generic, [see here](https://github.com/wellcometrust/wellcomecollection.org/issues/4294).

By creating forks, it feels like we would have to come back to this all the time, making it an unscalable bit of work stopping us from getting onto more interesting things.

File: content/webapp/pages/page.js (hunk 6 above)
Comment (jamesgorrie) [reply content/webapp/pages/page.js]: Could you have it as a `collectionVenue` slice, and then add a label of `closing` or something?

File: prismic-model/json/pages.json (hunk 5 above)
Comment (gestchild) [reply prismic-model/json/pages.json]: Think we might be talking at cross purposes, I'll come and chat

**Pull request merged**